docs/src/literate-tutorials/gpu_qp_heat_equation.jl

```diff
@@ -0,0 +1,254 @@
+    
+end
+
+function GPUGrid(grid::Grid{sdim}) where sdim
+    all_coords = cu(get_all_coordinates(grid))
+    n_cells = Int32(length(grid.cells))
+    GPUGrid(all_coords,n_cells)
+end
+
+
+struct GPUDofHandler{CDOFS<:AbstractArray{<:Number,1},GRID<:GPUGrid}<: Ferrite.AbstractDofHandler
+    cell_dofs::CDOFS
+    grid::GRID
+end
+
+
+function GPUDofHandler(dh::DofHandler)
+    GPUDofHandler(cu(Int32.(dh.cell_dofs)),GPUGrid(dh.grid))
+end
+
+function Adapt.adapt_structure(to, grid::GPUGrid)
+    all_coords = Adapt.adapt_structure(to, grid.all_coords)
+    n_cells = Adapt.adapt_structure(to, grid.n_cells)
+    GPUGrid(all_coords, n_cells)
+end
+
+function Adapt.adapt_structure(to, dh::GPUDofHandler)
+    cell_dofs = Adapt.adapt_structure(to, dh.cell_dofs)
+    grid = Adapt.adapt_structure(to, dh.grid)
+    GPUDofHandler(cell_dofs, grid)
+end
+
+function Adapt.adapt_structure(to, cv::Ferrite.StaticCellValues)
+    fv = Adapt.adapt_structure(to, cv.fv)
+    gm = Adapt.adapt_structure(to, cv.gm)
+    x = Adapt.adapt_structure(to, cu(cv.x))
+    weights = Adapt.adapt_structure(to, cv.weights)
+    Ferrite.StaticCellValues(fv,gm,x, weights)
+end
+
+
+gm = static_cellvalues.gm
+x = get_all_coordinates(grid)
+J = gm.dNdξ[1,1] ⊗ x[1] + gm.dNdξ[2,1] ⊗ x[2] + gm.dNdξ[3,1] ⊗ x[3] + gm.dNdξ[4,1] ⊗ x[4]
+det(J)
+inv_J = inv(J)
+inv_J ⋅ static_cellvalues.fv.dNdξ[1,1]
+
+function getjacobian(gm::Ferrite.StaticInterpolationValues,x,qr)
+    n_basefuncs = size(gm.Nξ,1)
+    J = gm.dNdξ[1,qr] ⊗ x[1]
+    for i = 2:n_basefuncs 
+        J+= gm.dNdξ[i,qr] ⊗ x[i]
+    end
+    return J
+end
+
+function assemble_element_gpu!(Kgpu,cv::Ferrite.StaticCellValues,dh::GPUDofHandler) 
+    tx = threadIdx().x 
+    bx = blockIdx().x
+    bd = blockDim().x
+    e = tx + (bx-1)*bd
+    n_cells = dh.grid.n_cells
+    e ≤ n_cells || return nothing # e here is the current element index.
+    n_qr = length(cv.weights)
+    n_basefuncs = size(cv.fv.Nξ,1)
+    dofs = dh.cell_dofs
+    x = dh.grid.all_coords
+    for qr = 1:n_qr # loop over quadrature points # TODO: propogate_ibounds
+        si = (e-1)*n_basefuncs
+        #J = gm.dNdξ[1,qr] ⊗ x[si+1] + gm.dNdξ[2,qr] ⊗ x[si+2] + gm.dNdξ[3,qr] ⊗ x[si+3] + gm.dNdξ[4,qr] ⊗ x[si+4]
+        cell_x = @view x[si+1:si+size(cv.gm.Nξ,1)]
+        J = getjacobian(cv.gm, cell_x,qr)
+        inv_J = inv(J)
+        #@cushow det(J)
+        @inbounds dΩ = det(J) * cv.weights[qr] 
+        for i = 1:n_basefuncs
+          @inbounds  ∇δu = inv_J ⋅ cv.fv.dNdξ[i,qr]
+            for j = 1:n_basefuncs
+               @inbounds ∇u = inv_J ⋅ cv.fv.dNdξ[j,qr]
+               @inbounds ig = dofs[(e-1)*n_basefuncs+i]
+               @inbounds jg = dofs[(e-1)*n_basefuncs+j]
+               CUDA.@atomic  Kgpu[ig, jg] += (∇δu ⋅ ∇u) * dΩ # atomic because many threads might write into the same memory addrres at the same time. 
+            end
+        end
+    end
+    return nothing
+end
+
+Kgpu = CUDA.zeros(dh.ndofs.x,dh.ndofs.x)
+gpu_dh = GPUDofHandler(dh)
```

Comment:
Maybe that's the plan anyway but wouldn't it be nicer to write `Adapt` rules for `DofHandler` and `Grid` which return `adapt_structure` for the GPU structs?

```julia
function Adapt.adapt_structure(to,dh:DofHandler)
    return adapt_structure(GPUDofHandler(cu(Int32.(dh.cell_dofs)),GPUGrid(dh.grid)))
end
```

and then just use the "normal" structs from a user perspective that are automatically converted when needed

Reply:
I am still a bit split about this, because it contains a performance pitfalls. If we have repeated assembly, then the dof handler will be converted and copied to GPU for each assembly instead of only once.

Reply:
Am I missing something or wouldn't the adapt call not also happen for each assembly kernel launch for the `GPUDofHandler` in that case?

Reply:
This should not happen, because we do not need to adapt the GPUDofHandler (it is already a GPU datastructure).